src/Store.cs

```diff
@@ -28,6 +28,32 @@ internal void AddFuel(ulong fuel)
             }
         }
 
+        internal object? GetData()
+        {
+            var data = Native.wasmtime_context_get_data(handle);
+            if (data == IntPtr.Zero)
+            {
+                return null;
+            }
+
+            return GCHandle.FromIntPtr(data).Target;
+        }
+
+        internal void SetData(object? data)
+        {
+            var oldData = Native.wasmtime_context_get_data(handle);
+            if (oldData != IntPtr.Zero) 
+            {
+                GCHandle.FromIntPtr(oldData).Free();
+            }
+            
+            if (data != null)
+            {
+                var newPtr = (IntPtr)GCHandle.Alloc(data);
+                Native.wasmtime_context_set_data(handle, newPtr);
+            }
```

Comment:
The call to `wasmtime_context_set_data` needs to be made unconditionally, to ensure we set the store data actually to `null`.
Otherwise, when calling `SetData(null)` and previously there was an object set, we would free the `GCHandle` here, but then the finalizer would attempt to free it again when disposing of the `Store` later (which can cause undefined behavior as the handle is no longer valid).

```suggestion
            var newPtr = IntPtr.Zero;
            if (data != null)
            {
                var newPtr = (IntPtr)GCHandle.Alloc(data);
            }            

            Native.wasmtime_context_set_data(handle, newPtr);
```

Reply:
Additionally, to be on the safe side if allocating a new `GCHandle` (or calling `wasmtime_context_set_data`) should ever fail, I think it might be a good idea to free the old handle only after calling `wasmtime_context_set_data` succeeded, i.e. move 
```cs
            if (oldData != IntPtr.Zero) 
            {
                GCHandle.FromIntPtr(oldData).Free();
            }
```
after the call to `wasmtime_context_set_data`, to prevent the situation that we free the `GCHandle` but it's still set as store data, so that the finalizer would then free it again.
Thanks!

Reply:
Done! Thanks a lot for being patient and helping me figure out a bunch of new concepts, your explanations were very helpful.